docs/design/automate-skus.md

```diff
@@ -0,0 +1,124 @@
+
+### Example locations output
+
+```shell
+$ aks-engine get-locations --client-id=${AZURE_CLIENT_ID} --client-secret=${AZURE_CLIENT_SECRET}
+Location            Name                      Latitude    Longitude
+australiacentral    Australia Central         -35.3075    149.1244
+australiacentral2   Australia Central 2       -35.3075    149.1244
+centraluseuap       Central US EUAP (Canary)  N/A         N/A
+chinaeast           China East                N/A         N/A
+...
+$ aks-engine get-locations -o code
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package helpers
+
+// AUTOGENERATED FILE
+
+// GetAzureLocations provides all azure regions in prod.
+// Related powershell to refresh this list:
+//   Get-AzureRmLocation | Select-Object -Property Location
+func GetAzureLocations() []string {
+	return []string{
+		"australiacentral",
+		"australiacentral2",
+		"centraluseuap",
+		"chinaeast",
+...
+
+```
+
+### Example SKUs output
+
+```shell
+$ aks-engine get-skus -o human
+Name                 Storage Account Type  Accelerated Networking Support
+Standard_A0          Standard_LRS          true
+Standard_A1          Standard_LRS          true
+Standard_D8s_v3      Premium_LRS           true
+Standard_DC1s_v2     Premium_LRS           false
+...
+$ aks-engine get-skus -o code
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package helpers
+
+// AUTOGENERATED FILE
+
+// GetKubernetesAllowedVMSKUs returns the allowed sizes for Kubernetes agent
+func GetKubernetesAllowedVMSKUs() string {
+	return `      "allowedValues": [
+		"Standard_A0",
+		"Standard_A1",
+		"Standard_D8s_v3",
+		"Standard_DC1s_v2",
+...
+```
+
+First we refactor the code so the data structures are in separate files. Then each file can be overwritten by redirecting the output of an `aks-engine` command.
+
+### Pros
+
+- No change to existing data structures, just a new way to refresh them
+- Keeps all knowledge of these Azure details in AKS Engine
+- Modest implementation effort
+
+### Cons
+
+- Hidden developer-facing commands aren't the best UX
+- Still requires a human to oversee the process
+
+## Alternative implementations considered
+
+### Dynamic API lookups
+
+Instead of consulting a lookup table in code, AKS Engine could make Azure API calls just-in-time. The location and SKU-related details of a cluster could be validated in real time by `aks-engine generate` or `aks-engine deploy` against current Azure information.
+
+#### Pros
+
+- Location and SKU support data is always current
+- Live Azure as the one source of truth
+
+#### Cons
+
+- Requires significant refactoring and implementation effort
+- Making API calls during `aks-engine generate` may be unwanted behavior
+- Somewhat slower and less reliable than using the data in code
```

Comment:
I agree that the most concrete downside is absorbing runtime API flakes for all aks-engine ARM operations. Removing that contingency is worth the additional "staleness" of the consts that we statically maintain in code, especially if we reduce the stale time by automating much of the process of updating those static consts.